skmorf/tree/_unsup_criterion.pyx

```diff
@@ -45,7 +77,7 @@ cdef class UnsupervisedCriterion(BaseCriterion):
         pass
```

Comment:
```suggestion
        self.sample_weight = sample_weight
        self.samples = samples
        self.start = start
        self.end = end
        self.n_node_samples = end - start
        self.weighted_n_samples = weighted_n_samples
        self.weighted_n_node_samples = 0.0

        cdef SIZE_t i
        cdef SIZE_t p
        cdef DOUBLE_t w = 1.0
        self.sum_total = 0.0

        for p in range(start, end):
            i = samples[p]

            # w is originally set to be 1.0, meaning that if no sample weights
            # are given, the default weight of each sample is 1.0
            if sample_weight != NULL:
                w = sample_weight[i]
            
            # ith sample; overall compute the weighted average
            self.sum_total += w * X[i]
            self.weighted_n_node_samples += w

        # Reset to pos=start
        self.reset()
        return 0
```

They should share the `init(...)` function. I mentioned that all subclasses most likely share this same interface. If you skim the formulas for fastBIC and TwoMeans, there is no real extra data components (that I'm aware of).